cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,299 @@
+		clusterName:   opts.ClusterName,
+		heatClient:    heatClient,
+		waitTimeStep:  waitForStatusTimeStep,
+	}
+
+	manager.stackID, err = manager.GetStackID()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack ID on manager: %v", err)
+	}
+	manager.stackName, err = manager.GetStackName()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack name on manager: %v", err)
+	}
+
+	return &manager, nil
+}
+
+// NodeGroupSize gets the current cluster size as reported by magnum.
+// The nodegroup argument is ignored as this implementation of OpenstackManager
+// assumes that only a single node group exists.
+func (osm *OpenstackManagerHeat) NodeGroupSize(nodegroup string) (int, error) {
+	cluster, err := clusters.Get(osm.clusterClient, osm.clusterName).Extract()
+	if err != nil {
+		return 0, fmt.Errorf("could not get cluster: %v", err)
+	}
+	return cluster.NodeCount, nil
+}
+
+// UpdateNodeCount replaces the cluster node_count in magnum.
+func (osm *OpenstackManagerHeat) UpdateNodeCount(nodegroup string, nodes int) error {
+	updateOpts := []clusters.UpdateOptsBuilder{
+		UpdateOptsInt{Op: clusters.ReplaceOp, Path: "/node_count", Value: nodes},
+	}
+	_, err := clusters.Update(osm.clusterClient, osm.clusterName, updateOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("could not update cluster: %v", err)
+	}
+	return nil
+}
+
+// GetNodes currently returns nothing and this seems to work fine.
+func (osm *OpenstackManagerHeat) GetNodes(nodegroup string) ([]string, error) {
+	// TODO: get nodes from heat?
+	// This works fine being empty for now anyway
+	return []string{}, nil
+}
+
+// DeleteNodes deletes nodes by passing a comma separated list of names or IPs
+// of minions to remove to heat, and simultaneously sets the new number of minions on the stack.
+// The magnum node_count is then set to the new value (does not cause any more nodes to be removed).
+func (osm *OpenstackManagerHeat) DeleteNodes(nodegroup string, minionsToRemove []string, updatedNodeCount int) error {
+	minionsList := strings.Join(minionsToRemove, ",")
+	updateOpts := stacks.UpdateOpts{
+		Parameters: map[string]interface{}{
+			"minions_to_remove": minionsList,
+			"number_of_minions": updatedNodeCount,
+		},
+	}
+
+	updateResult := stacks.UpdatePatch(osm.heatClient, osm.stackName, osm.stackID, updateOpts)
+	err := updateResult.ExtractErr()
+	if err != nil {
+		return fmt.Errorf("stack patch failed: %v", err)
+	}
+
+	// Wait for the stack to do its thing before updating the cluster node_count
+	err = osm.WaitForStackStatus(stackStatusUpdateInProgress, waitForUpdateStatusTimeout)
+	if err != nil {
+		return fmt.Errorf("error waiting for stack %s status: %v", stackStatusUpdateInProgress, err)
+	}
+	err = osm.WaitForStackStatus(stackStatusUpdateComplete, waitForCompleteStatusTimout)
+	if err != nil {
+		return fmt.Errorf("error waiting for stack %s status: %v", stackStatusUpdateComplete, err)
+	}
+
+	err = osm.UpdateNodeCount(nodegroup, updatedNodeCount)
```

Comment:
Nodes are deleted through Heat and then Magnum is updated afterward? The wait-for-update-to-finish makes a bit more sense now :). I hope that we could go all through Heat or all through Magnum. I know that the scale down in Magnum does not currently support node specification, so while that is being worked on, maybe Heat could be used exclusively for scale up and scale down and have Magnum update itself?

(sorry this comment was previously left on the wrong line of the wrong file)

Reply:
+1 it seems like a poor pattern to upstream code on Magnums shortcomings but rather just Heat for now only until Magnum can support the operation, then it makes sense to use Magnum only.

Reply:
@scott-chaney 
Yes, the call to UpdateNodeCount is purely to set the node_count in magnum to the correct value, without actually deleting any more nodes. I would prefer that all interactions were through magnum but to delete the correct nodes on scale down it is necessary to do this one step with heat.

It would be possible to reverse things and use heat as much as possible, but then that requires doing multiple steps with magnum, to update its node count both on the scale up and scale down.

> maybe Heat could be used exclusively for scale up and scale down and have Magnum update itself?

How would magnum update itself?

Reply:
Maybe the direction of this PR, as it is or wants to be, needs to be dependent on Magnum adding the functionality first, otherwise this split behavior between Magnum up and Heat down is an anti-pattern. Since Magnum does not have this now, and no work seems to be scheduled to provide it, such as Magnum-API handling updates to the heat stacks that are not database updates for Magnum, currently upstream-able content should focus on a Heat-only based solution until Magnum can properly support the change. We'd like to further collaborate more about the Magnum-level solution offline as well =)

Reply:
> How would magnum update itself?

I'm not sure what the best method would be from Magnum's standpoint but I was thinking something like whenever the cluster is queried or the size parameter was updated, Magnum could first check the stack and update if needed.

Reply:
overall, I think workding towards finding a way to reduce the overall state duplication between the two would be optimal

Reply:
There are two main things Magnum needs and should add for this implementation to be cleaner: node groups with min/max size, and the ability to target specific node deletion when scaling down. The interactions and APIs should be very similar to what GKE offers.

Both are being worked on and maybe pushing a first version that temporary calls heat in a couple places is not too bad? In the end we will likely need two openstack implementations of the cloud provider, the same way there's gce and gke.

Reply:
> Both are being worked on and maybe pushing a first version that temporary calls heat in a couple places is not too bad?

The mixture of API calls in unfortunate, but I think secondary to the concern about essentially issuing two separate scaling operations.  The stack status polling seems to be implemented in part to ensure that the magnum update never preempts the Heat operations, and I am hoping that we can rework that polling strategy for efficiency.

Reply:
Hey Ricardo!  o/ 
> Both are being worked on and maybe pushing a first version that temporary calls heat in a couple places is not too bad?

No not at all, this is perfectly reasonable. Just wanted to push some bigger ideas out, especially regarding the ability to reduce Magnums complexity/scope and state duplication in favor of cleaner, more accurate patterns. 
BTW, lets all connect again soon. Maybe some skyping is in order =-D





>In the end we will likely need two openstack implementations of the cloud provider, the same way there's gce and gke.

Yeah i think @ricolin  has a heat specific version